proposals/p2173.md

```diff
@@ -0,0 +1,489 @@
+  let Element:! Type;
+  let Slice:! Container where .Element = Element;
+}
+// `T.Slice.Element` rewritten to `T.Element` because type of `T.Slice` says `.Element = Element`.
+// `T.Element` rewritten to `i32` because type of `T` says `.Element = i32`.
+fn A[T:! Container where .Element = i32](x: T.Slice.Element) {}
+```
+
+Rewrites aren't performed on the left-hand side of such an `=`, so
+`where .A = .B and .A = C` is not rewritten to `where .A = .B and .B = C`.
+Instead, such a `where` clause is invalid unless each rule for `.A` specifies
+the same rewrite.
+
+When an `=` appears within an interface, it is possible to form a cycle where a
+type becomes more complex with each iteration. In order to avoid this problem,
+if, when substituting into the right-hand side of such an `=`, the same rewrite
+is performed again, the program is invalid.
+
+```
+interface Container {
+  let Element:! Type;
+  // (1), bad constraint
+  let Slice:! Container where .Element = .Slice.Element;
+}
+// Error, `T.Slice.Element` rewritten to
+// `T.Slice.Slice.Element` by (1), then to
+// `T.Slice.Slice.Slice.Element` by (1)
+fn A[T:! Container](x: T.Slice.Element) {}
+```
+
+It is valid for the same rewrite rule to be used multiple times within a type
+expression, so long as it is not used during substitution into its own
+right-hand side:
+
+```
+interface BInterface {
+  let C:! Type;
+}
+interface AInterface {
+  let D:! AInterface;
+  let B:! BInterface where .C = Self.D;
+}
+// `A.B.C` is rewritten to `Self.D`, because the type of `A.B` says `.C = Self.D`.
+//     `Self`=`A` is substituted into this giving `A.D`.
+//     During this substitution, the same rewrite rule is not used again,
+//     so this is OK.
+// Then `A.D.B.C` is rewritten to `Self.D`, because the type of `A.D.B` says
+// `.C = Self.D`.
+//     `Self`=`A.D` is substituted into this giving `A.D.D`.
+//     During this substitution, the same rewrite rule is not used again,
+//     so this is also OK, even though the same rule was used twice
+//     in forming the overall type.
+fn F(A:! AInterface, d: A.B.C.B.C) {}
+```
+
+Rewrite constraints thereby give us a deterministic, terminating type
+canonicalization mechanism for associated constants: in `A.B`, if the type of
+`A` specifies that `.B == C`, then `A.B` is replaced by `C`. Equality of types
+constrained in this way is transitive.
+
+Note that given `T:! C where .R = i32` can result in a type `T.R` whose behavior
+is different from the behavior of `T.R` given `T:! C`. For example, member
+lookup into `T.R` can find different results and operations can therefore have
+different behavior. However, this does not violate coherence because `C` and
+`C where .R = i32` don't differ by merely having more type information; rather,
+they are different types that have an isomorphic set of values, somewhat like
+`i32` and `u32`. An `=` constraint is not merely learning a new fact about a
+type, it is requesting different behavior.
+
```

Comment:
Would it be useful to have some dedicated sections with examples to clarify the rules? Here are some that I think illustrate some important points:

A variation on the example above:
```
interface Allowed1 {
  let A:! Allowed1 where .A = .Self;
}
fn F1(T:! Allowed1, x: T.A.A.A);
// Allowed, since rewrite is only applied once per sub-expression
// T.A.A.A = (T.A.A).A -> (T.A).A = T.A.A -> T.A
```

`Allowed1` with the sides switched leads to errors (similar error as your second `Container` example).
```
interface Error1 {
  let E:! Error1 where .Self = .E;
}
fn F2(T:! Error1, x: T.E);
// T.E -> T.E.E -> T.E.E.E error, rewrite applied twice.
// Clearly can't keep applying this rewrite.
// Heuristic: put the "shorter" expression on the right.
```
This is a case we could diagnose with just the interface definition, since the right-hand side of the = already matches the rewrite rule.

Two rewrites create a loop:
```
interface Example2 {
  let B:! Example2;
  let C:! Example2 where .Self = B;
}
// `where .Self = ...` by itself is fine. `
fn AllowedEx2(T:! Error2, x: T.C);
// But combined with another rewrite, creates an infinite loop.
fn ErrorEx2(T:! Error2 where .B = .C, x: T.C);
// T.C -> T.B -> T.C -> T.B, error: applies T.C -> T.B twice.
// Using `==` instead of `=` would make this redundant, not an error.
```

Rewrite applies twice, but doesn't create an infinite loop
```
constraint ForwardDeclaredConstraint;
interface Error3 {
  let X:! Error3;
  // Means `Y:! Error3 where .X = X.Y`
  let Y:! ForwardDeclaredConstraint;
}
constraint ForwardDeclaredConstraint {
  extends Error3 where .X = X.Y;
}
fn F4[(T:! Error3, x: T.Y.Y.X);
// T.Y.Y.X -> T.Y.X.Y -> T.X.Y.Y
// An error since it applies the same rewrite twice, even
// though there is no infinite loop.
```
This is in some sense a "false positive" of our policy, and the rules could be modified to allow this if this comes up in real code.

Reply:
```
interface Helper2 {
  let D:! Type;
}
interface Example2 {
  let B:! Helper2;
  let C:! Helper2 where .D = B.D;
}
// `where .Self = ...` by itself is fine. `
fn AllowedEx2(T:! Error2, x: T.C.D);
// But combined with another rewrite, creates an infinite loop.
fn ErrorEx2(T:! Error2 where .B.D = .C.D, x: T.C.D);
// T.C.D -> T.B.D -> T.C.D -> T.B.D, error: applies T.C.D -> T.B.D twice.
// Using `==` instead of `=` would make this redundant, not an error.
```

Reply:
Rewriting `Error1` to avoid `where .Self =`:
```